File: python/questions/maxProduct1.py
from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ln = len(nums)
        dp = [nums[0]] * 2
        ans = nums[0]
        for i in range(1, ln):
            n = nums[i]
            l, h = dp
            dp[0] = min(l*n, h*n, n) 
            dp[1] = max(l*n, h*n, n)
            ans = max(ans, dp[1])
        return ans

File: cpp/questions/maxProduct1.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        std::pair<int, int> dp = {nums[0], nums[0]};
        int ans = nums[0];
        for (int i=1; i<nums.size(); ++i) {
            int n = nums[i];
            auto [l, h] = dp;
            dp.first = std::min(std::min(l*n, h*n), n);
            dp.second = std::max(std::max(l*n, h*n), n);
            ans = std::max(ans, dp.second);
        }
        return ans;
    }
};
